src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp

```diff
@@ -0,0 +1,252 @@
+	std::list<NodePtr> frequenciesList;
+	std::copy(std::make_move_iterator(activeSymbols),
+		std::make_move_iterator(frequencies.end()),
+		std::back_inserter(frequenciesList));
+
+	// If there are no elements, return an empty node
+	if(frequenciesList.empty()) {
+		return nullptr;
+	}
+
+	// While there is more than one element in the list
+	while(frequenciesList.size() > 1) {
+		// Pop first 2 elements
+		auto n1 = std::move(frequenciesList.front());
+		frequenciesList.pop_front();
+		auto n2 = std::move(frequenciesList.front());
+		frequenciesList.pop_front();
+		// Create a new one which is their sum
+		auto newNode = n1 + n2;
+		// Insert it to the correct location in the list
+		auto newLoc = std::upper_bound(frequenciesList.begin(),
+			frequenciesList.end(),
+			newNode->frequency,
+			[](int frequency, NodePtr &n) { return n->frequency > frequency; });
+		frequenciesList.insert(newLoc, std::move(newNode));
+	}
+
+	return std::move(frequenciesList.front());
+}
+
+std::pair<bool, HuffmanEncoderDecoder::MappingType>
+	HuffmanEncoderDecoder::createMapping(const NodePtr &root) {
+
+	MappingType mapping;
+
+	// If the tree is empty, then so should the mapping be
+	if(root == nullptr) {
+		return {false, mapping};
+	}
+
+	// If the tree has only one element, we want to assign it a single-bit 0
+	// code
+	if(root->leftSon == nullptr && root->rightSon == nullptr) {
+		mapping.insert(
+			root->character, std::make_pair<unsigned char, unsigned int>(1, 0));
+		return {true, mapping};
+	}
+
+	// Otherwise we use dfs on the tree with an initial code of 0 length
+	try {
+		HuffmanEncoderDecoder::dfs(mapping,
+			root,
+			std::make_pair<unsigned char, unsigned int>(0, 0),
+			32);
+	} catch(std::length_error) { return {false, mapping}; }
+
+	return {true, mapping};
+}
+
+void HuffmanEncoderDecoder::dfs(MappingType &mapping,
+	const NodePtr &n,
+	std::pair<unsigned char, unsigned int> code,
+	int depth) {
+
+	// We reached a leaf node, so we insert it to the mapping
+	if(n->leftSon == nullptr && n->rightSon == nullptr) {
+		mapping.insert(n->character, code);
+		return;
+	}
+
+	// If depth is 0, than we won't have enough space to represent certain
+	// characters
+	if(depth == 0) {
+		throw std::length_error(
+			"The tree depth is more than the max representation length");
+	}
+
+	// If we have a left son, we concatenate a 0 to the code, and recurse
+	if(n->leftSon != nullptr) {
+		code.first++;
+		code.second *= 2;
+		dfs(mapping, n->leftSon, code, depth - 1);
+		code.first--;
+		code.second /= 2;
+	}
+
+	// If we have a left son, we concatenate a 1 to the code, and recurse
+	if(n->rightSon != nullptr) {
+		code.first++;
+		code.second = (2 * code.second) + 1;
+		dfs(mapping, n->rightSon, code, depth - 1);
+		code.first--;
+		code.second = (code.second - 1) / 2;
+	}
+}
+
+std::pair<bool, HuffmanEncoderDecoder::EncodedType>
+	HuffmanEncoderDecoder::encode(const std::string &plaintext) const {
+
+	constexpr unsigned char BITS_IN_CHAR = 8;
```

Comment:
`std::numeric_limits<char>::digits`

Reply:
Cool, this is also relevant in a couple of other places.

TODO